numpy/linalg/linalg.py

```diff
@@ -2536,7 +2528,7 @@ def _multi_dot_matrix_chain_order(arrays, return_costs=False):
             j = i + l
             m[i, j] = Inf
             for k in range(i, j):
-                q = m[i, k] + m[k+1, j] + p[i]*p[k+1]*p[j+1]
+                q = m[i, k] + m[k+1, j] + p[i]*p[k+1]*p[j+1]*max(dim[i:j+1])
```

Comment:
I don't think this new expression `max(dim[i:j+1])` for calculating the number of repeats in broadcasting is correct.

Consider matrix multiplication between inputs of shape `(N, 1, X, Y)` and `(1, M, X, Y)`. These inputs would have `dim = [N, M]`, but the number of repeats per broadcasting rules is `N*M`, not `max(N, M)`.

Something like the utility function [`_broadcast_shape()`](https://github.com/numpy/numpy/blob/4c349d855c12e6f0ba0bf925939f7a8156c825d8/numpy/lib/stride_tricks.py#L176) from `stride_tricks` could potentially be helpful for calculating this.